base/utf16.jl

```diff
@@ -30,27 +107,29 @@ function endof(s::UTF16String)
     d = s.data
     i = length(d) - 1
     i == 0 && return i
-    utf16_is_surrogate(d[i]) ? i-1 : i
+    (utf16_is_surrogate(d[i]) ? i-1 : i)
 end
 
 function next(s::UTF16String, i::Int)
-    if !utf16_is_surrogate(s.data[i])
-        return Char(s.data[i]), i+1
-    elseif length(s.data)-1 > i && utf16_is_lead(s.data[i]) && utf16_is_trail(s.data[i+1])
-        return utf16_get_supplementary(s.data[i], s.data[i+1]), i+2
-    end
-    throw(ArgumentError("invalid UTF-16 character index"))
+    ch = s.data[i]
+    !utf16_is_surrogate(ch) && return (Char(ch), i+1)
+    # check length, account for terminating \0
+    i >= (length(s.data)-1) && utf_errfunc(UTF_ERR_MISSING_SURROGATE, i, UInt32(ch))
+    !utf16_is_lead(ch) && utf_errfunc(UTF_ERR_NOT_LEAD, i, ch)
+    ct = s.data[i+1]
+    !utf16_is_trail(ct) && utf_errfunc(UTF_ERR_NOT_TRAIL, i, ch)
+    utf16_get_supplementary(ch, ct), i+2
 end
 
 function reverseind(s::UTF16String, i::Integer)
     j = length(s.data) - i
-    return Base.utf16_is_trail(s.data[j]) ? j-1 : j
+    (Base.utf16_is_trail(s.data[j]) ? j-1 : j)
```

Comment:
Ah, that explains the strange errors I was getting that it couldn't find a method with "::Void" in the argument list... Thanks!